java/azure-eventhubs/src/main/java/com/microsoft/azure/servicebus/MessageReceiver.java

```diff
@@ -300,16 +300,16 @@ public void setReceiveTimeout(final Duration value)
 				@Override
 				public void onEvent()
 				{
-					if (receiveLink.getLocalState() == EndpointState.CLOSED || receiveLink.getRemoteState() == EndpointState.CLOSED)
-					{
-						createReceiveLink();
-					}
-
 					final List<Message> messages = receiveCore(maxMessageCount);
 					if (messages != null)
 						onReceive.complete(messages);
 					else
 						pendingReceives.offer(new ReceiveWorkItem(onReceive, receiveTimeout, maxMessageCount));
+
+					if (receiveLink.getLocalState() == EndpointState.CLOSED || receiveLink.getRemoteState() == EndpointState.CLOSED)
+					{
+						createReceiveLink();
+					}
```

Comment:
If this order of "put workItem into pendingQueue" --> "recreate Link if broken" actually matters in terms of race condition, I think we should have some base implementation that enforce this order everywhere so that sub-class implementation does not make this type of mistake easily (e.g. I am worry when we write the messaging implementation we would hit the same thing again).